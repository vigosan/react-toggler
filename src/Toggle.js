import { Component } from 'react';
import PropTypes from 'prop-types';

class Toggle extends Component {
  static propTypes = {
    children: PropTypes.func,
    defaultOn: PropTypes.bool,
    render: PropTypes.func,
    onToggle: PropTypes.func
  };

  static defaultProps = {
    children: () => null,
    defaultOn: false,
    onToggle: () => {},
    render: () => null
  };

  state = {
    on: this.props.defaultOn
  };

  toggle = () => {
    this.setState(
      prevState => {
        return { on: !prevState.on };
      },
      () => this.props.onToggle(this.state)
    );
  };

  getRenderProps() {
    const { onToggle } = this.props;

    return { onToggle, toggle: this.toggle, ...this.state };
  }

  render() {
    const { children, render } = this.props;
    const renderProps = this.getRenderProps();

    if (render) {
      return render(renderProps);
    }

    return children(renderProps);
  }
}

export default Toggle;
