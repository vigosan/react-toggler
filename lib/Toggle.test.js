import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import Toggle from './Toggle';

describe('Toggle', () => {
  let wrapper;
  let on;
  let toggle;
  let onToggle = jest.fn();

  beforeEach(() => {
    wrapper = mount(
      <Toggle
        onToggle={onToggle}
        render={({ on: tOn, toggle: tToggle }) => {
          on = tOn;
          toggle = tToggle;

          return null;
        }}
      />
    );
  });

  afterEach(function() {
    onToggle.mockReset();
  });

  it('renders its children', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('starts off', () => {
    expect(on).toBe(false);
  });

  it('toggles on', () => {
    toggle();

    expect(on).toBe(true);
  });

  it('executes a callback when toggled', () => {
    toggle();

    expect(onToggle).toBeCalledWith({
      on: true
    });
  });
});
