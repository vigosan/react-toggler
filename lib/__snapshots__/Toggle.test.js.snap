// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders its children 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Toggle
    onToggle={[Function]}
    render={[Function]}
  >
    [Function]
  </Toggle>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": Toggle {
      "_reactInternalFiber": FiberNode {
        "_debugID": 4,
        "_debugIsCurrentlyTiming": false,
        "_debugOwner": FiberNode {
          "_debugID": 3,
          "_debugIsCurrentlyTiming": false,
          "_debugOwner": null,
          "_debugSource": null,
          "alternate": null,
          "child": [Circular],
          "effectTag": 1,
          "firstEffect": null,
          "index": 0,
          "internalContextTag": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": Object {
            "Component": [Function],
            "context": null,
            "props": Object {
              "children": [Function],
              "onToggle": [Function],
              "render": [Function],
            },
          },
          "memoizedState": Object {
            "context": null,
            "mount": true,
            "props": Object {
              "children": [Function],
              "onToggle": [Function],
              "render": [Function],
            },
          },
          "nextEffect": null,
          "pendingProps": null,
          "pendingWorkPriority": 0,
          "ref": null,
          "return": FiberNode {
            "_debugID": 1,
            "_debugIsCurrentlyTiming": false,
            "_debugOwner": null,
            "_debugSource": null,
            "alternate": FiberNode {
              "_debugID": 1,
              "_debugIsCurrentlyTiming": false,
              "_debugOwner": null,
              "_debugSource": null,
              "alternate": [Circular],
              "child": null,
              "effectTag": 0,
              "firstEffect": null,
              "index": 0,
              "internalContextTag": 0,
              "key": null,
              "lastEffect": null,
              "memoizedProps": null,
              "memoizedState": null,
              "nextEffect": null,
              "pendingProps": null,
              "pendingWorkPriority": 1,
              "ref": null,
              "return": null,
              "sibling": null,
              "stateNode": Object {
                "containerInfo": <div />,
                "context": Object {},
                "current": [Circular],
                "isScheduled": false,
                "nextScheduledRoot": null,
                "pendingContext": null,
              },
              "tag": 3,
              "type": null,
              "updateQueue": Object {
                "callbackList": null,
                "first": Object {
                  "callback": null,
                  "isForced": false,
                  "isReplace": false,
                  "isTopLevelUnmount": false,
                  "next": null,
                  "partialState": Object {
                    "element": <WrapperComponent
                      Component={[Function]}
                      context={null}
                      props={
                        Object {
                          "children": [Function],
                          "onToggle": [Function],
                          "render": [Function],
                        }
                      }
                    />,
                  },
                  "priorityLevel": 1,
                },
                "hasForceUpdate": false,
                "isProcessing": false,
                "last": Object {
                  "callback": null,
                  "isForced": false,
                  "isReplace": false,
                  "isTopLevelUnmount": false,
                  "next": null,
                  "partialState": Object {
                    "element": <WrapperComponent
                      Component={[Function]}
                      context={null}
                      props={
                        Object {
                          "children": [Function],
                          "onToggle": [Function],
                          "render": [Function],
                        }
                      }
                    />,
                  },
                  "priorityLevel": 1,
                },
              },
            },
            "child": [Circular],
            "effectTag": 0,
            "firstEffect": null,
            "index": 0,
            "internalContextTag": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": null,
            "memoizedState": Object {
              "element": <WrapperComponent
                Component={[Function]}
                context={null}
                props={
                  Object {
                    "children": [Function],
                    "onToggle": [Function],
                    "render": [Function],
                  }
                }
              />,
            },
            "nextEffect": null,
            "pendingProps": null,
            "pendingWorkPriority": 0,
            "ref": null,
            "return": null,
            "sibling": null,
            "stateNode": Object {
              "containerInfo": <div />,
              "context": Object {},
              "current": [Circular],
              "isScheduled": false,
              "nextScheduledRoot": null,
              "pendingContext": null,
            },
            "tag": 3,
            "type": null,
            "updateQueue": null,
          },
          "sibling": null,
          "stateNode": WrapperComponent {
            "_reactInternalFiber": [Circular],
            "_reactInternalInstance": Object {},
            "context": Object {},
            "props": Object {
              "Component": [Function],
              "context": null,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "refs": Object {},
            "state": Object {
              "context": null,
              "mount": true,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
          "tag": 2,
          "type": [Function],
          "updateQueue": null,
        },
        "_debugSource": null,
        "alternate": null,
        "child": null,
        "effectTag": 1,
        "firstEffect": null,
        "index": 0,
        "internalContextTag": 0,
        "key": null,
        "lastEffect": null,
        "memoizedProps": Object {
          "children": [Function],
          "onToggle": [Function],
          "render": [Function],
        },
        "memoizedState": Object {
          "on": false,
        },
        "nextEffect": null,
        "pendingProps": null,
        "pendingWorkPriority": 0,
        "ref": null,
        "return": FiberNode {
          "_debugID": 3,
          "_debugIsCurrentlyTiming": false,
          "_debugOwner": null,
          "_debugSource": null,
          "alternate": null,
          "child": [Circular],
          "effectTag": 1,
          "firstEffect": null,
          "index": 0,
          "internalContextTag": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": Object {
            "Component": [Function],
            "context": null,
            "props": Object {
              "children": [Function],
              "onToggle": [Function],
              "render": [Function],
            },
          },
          "memoizedState": Object {
            "context": null,
            "mount": true,
            "props": Object {
              "children": [Function],
              "onToggle": [Function],
              "render": [Function],
            },
          },
          "nextEffect": null,
          "pendingProps": null,
          "pendingWorkPriority": 0,
          "ref": null,
          "return": FiberNode {
            "_debugID": 1,
            "_debugIsCurrentlyTiming": false,
            "_debugOwner": null,
            "_debugSource": null,
            "alternate": FiberNode {
              "_debugID": 1,
              "_debugIsCurrentlyTiming": false,
              "_debugOwner": null,
              "_debugSource": null,
              "alternate": [Circular],
              "child": null,
              "effectTag": 0,
              "firstEffect": null,
              "index": 0,
              "internalContextTag": 0,
              "key": null,
              "lastEffect": null,
              "memoizedProps": null,
              "memoizedState": null,
              "nextEffect": null,
              "pendingProps": null,
              "pendingWorkPriority": 1,
              "ref": null,
              "return": null,
              "sibling": null,
              "stateNode": Object {
                "containerInfo": <div />,
                "context": Object {},
                "current": [Circular],
                "isScheduled": false,
                "nextScheduledRoot": null,
                "pendingContext": null,
              },
              "tag": 3,
              "type": null,
              "updateQueue": Object {
                "callbackList": null,
                "first": Object {
                  "callback": null,
                  "isForced": false,
                  "isReplace": false,
                  "isTopLevelUnmount": false,
                  "next": null,
                  "partialState": Object {
                    "element": <WrapperComponent
                      Component={[Function]}
                      context={null}
                      props={
                        Object {
                          "children": [Function],
                          "onToggle": [Function],
                          "render": [Function],
                        }
                      }
                    />,
                  },
                  "priorityLevel": 1,
                },
                "hasForceUpdate": false,
                "isProcessing": false,
                "last": Object {
                  "callback": null,
                  "isForced": false,
                  "isReplace": false,
                  "isTopLevelUnmount": false,
                  "next": null,
                  "partialState": Object {
                    "element": <WrapperComponent
                      Component={[Function]}
                      context={null}
                      props={
                        Object {
                          "children": [Function],
                          "onToggle": [Function],
                          "render": [Function],
                        }
                      }
                    />,
                  },
                  "priorityLevel": 1,
                },
              },
            },
            "child": [Circular],
            "effectTag": 0,
            "firstEffect": null,
            "index": 0,
            "internalContextTag": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": null,
            "memoizedState": Object {
              "element": <WrapperComponent
                Component={[Function]}
                context={null}
                props={
                  Object {
                    "children": [Function],
                    "onToggle": [Function],
                    "render": [Function],
                  }
                }
              />,
            },
            "nextEffect": null,
            "pendingProps": null,
            "pendingWorkPriority": 0,
            "ref": null,
            "return": null,
            "sibling": null,
            "stateNode": Object {
              "containerInfo": <div />,
              "context": Object {},
              "current": [Circular],
              "isScheduled": false,
              "nextScheduledRoot": null,
              "pendingContext": null,
            },
            "tag": 3,
            "type": null,
            "updateQueue": null,
          },
          "sibling": null,
          "stateNode": WrapperComponent {
            "_reactInternalFiber": [Circular],
            "_reactInternalInstance": Object {},
            "context": Object {},
            "props": Object {
              "Component": [Function],
              "context": null,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "refs": Object {},
            "state": Object {
              "context": null,
              "mount": true,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "updater": Object {
              "enqueueForceUpdate": [Function],
              "enqueueReplaceState": [Function],
              "enqueueSetState": [Function],
              "isMounted": [Function],
            },
          },
          "tag": 2,
          "type": [Function],
          "updateQueue": null,
        },
        "sibling": null,
        "stateNode": [Circular],
        "tag": 2,
        "type": [Function],
        "updateQueue": null,
      },
      "_reactInternalInstance": Object {},
      "context": Object {},
      "props": Object {
        "children": [Function],
        "onToggle": [Function],
        "render": [Function],
      },
      "refs": Object {},
      "state": Object {
        "on": false,
      },
      "toggle": [Function],
      "updater": Object {
        "enqueueForceUpdate": [Function],
        "enqueueReplaceState": [Function],
        "enqueueSetState": [Function],
        "isMounted": [Function],
      },
    },
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": [Function],
      "onToggle": [Function],
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": Toggle {
        "_reactInternalFiber": FiberNode {
          "_debugID": 4,
          "_debugIsCurrentlyTiming": false,
          "_debugOwner": FiberNode {
            "_debugID": 3,
            "_debugIsCurrentlyTiming": false,
            "_debugOwner": null,
            "_debugSource": null,
            "alternate": null,
            "child": [Circular],
            "effectTag": 1,
            "firstEffect": null,
            "index": 0,
            "internalContextTag": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": Object {
              "Component": [Function],
              "context": null,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "memoizedState": Object {
              "context": null,
              "mount": true,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "nextEffect": null,
            "pendingProps": null,
            "pendingWorkPriority": 0,
            "ref": null,
            "return": FiberNode {
              "_debugID": 1,
              "_debugIsCurrentlyTiming": false,
              "_debugOwner": null,
              "_debugSource": null,
              "alternate": FiberNode {
                "_debugID": 1,
                "_debugIsCurrentlyTiming": false,
                "_debugOwner": null,
                "_debugSource": null,
                "alternate": [Circular],
                "child": null,
                "effectTag": 0,
                "firstEffect": null,
                "index": 0,
                "internalContextTag": 0,
                "key": null,
                "lastEffect": null,
                "memoizedProps": null,
                "memoizedState": null,
                "nextEffect": null,
                "pendingProps": null,
                "pendingWorkPriority": 1,
                "ref": null,
                "return": null,
                "sibling": null,
                "stateNode": Object {
                  "containerInfo": <div />,
                  "context": Object {},
                  "current": [Circular],
                  "isScheduled": false,
                  "nextScheduledRoot": null,
                  "pendingContext": null,
                },
                "tag": 3,
                "type": null,
                "updateQueue": Object {
                  "callbackList": null,
                  "first": Object {
                    "callback": null,
                    "isForced": false,
                    "isReplace": false,
                    "isTopLevelUnmount": false,
                    "next": null,
                    "partialState": Object {
                      "element": <WrapperComponent
                        Component={[Function]}
                        context={null}
                        props={
                          Object {
                            "children": [Function],
                            "onToggle": [Function],
                            "render": [Function],
                          }
                        }
                      />,
                    },
                    "priorityLevel": 1,
                  },
                  "hasForceUpdate": false,
                  "isProcessing": false,
                  "last": Object {
                    "callback": null,
                    "isForced": false,
                    "isReplace": false,
                    "isTopLevelUnmount": false,
                    "next": null,
                    "partialState": Object {
                      "element": <WrapperComponent
                        Component={[Function]}
                        context={null}
                        props={
                          Object {
                            "children": [Function],
                            "onToggle": [Function],
                            "render": [Function],
                          }
                        }
                      />,
                    },
                    "priorityLevel": 1,
                  },
                },
              },
              "child": [Circular],
              "effectTag": 0,
              "firstEffect": null,
              "index": 0,
              "internalContextTag": 0,
              "key": null,
              "lastEffect": null,
              "memoizedProps": null,
              "memoizedState": Object {
                "element": <WrapperComponent
                  Component={[Function]}
                  context={null}
                  props={
                    Object {
                      "children": [Function],
                      "onToggle": [Function],
                      "render": [Function],
                    }
                  }
                />,
              },
              "nextEffect": null,
              "pendingProps": null,
              "pendingWorkPriority": 0,
              "ref": null,
              "return": null,
              "sibling": null,
              "stateNode": Object {
                "containerInfo": <div />,
                "context": Object {},
                "current": [Circular],
                "isScheduled": false,
                "nextScheduledRoot": null,
                "pendingContext": null,
              },
              "tag": 3,
              "type": null,
              "updateQueue": null,
            },
            "sibling": null,
            "stateNode": WrapperComponent {
              "_reactInternalFiber": [Circular],
              "_reactInternalInstance": Object {},
              "context": Object {},
              "props": Object {
                "Component": [Function],
                "context": null,
                "props": Object {
                  "children": [Function],
                  "onToggle": [Function],
                  "render": [Function],
                },
              },
              "refs": Object {},
              "state": Object {
                "context": null,
                "mount": true,
                "props": Object {
                  "children": [Function],
                  "onToggle": [Function],
                  "render": [Function],
                },
              },
              "updater": Object {
                "enqueueForceUpdate": [Function],
                "enqueueReplaceState": [Function],
                "enqueueSetState": [Function],
                "isMounted": [Function],
              },
            },
            "tag": 2,
            "type": [Function],
            "updateQueue": null,
          },
          "_debugSource": null,
          "alternate": null,
          "child": null,
          "effectTag": 1,
          "firstEffect": null,
          "index": 0,
          "internalContextTag": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": Object {
            "children": [Function],
            "onToggle": [Function],
            "render": [Function],
          },
          "memoizedState": Object {
            "on": false,
          },
          "nextEffect": null,
          "pendingProps": null,
          "pendingWorkPriority": 0,
          "ref": null,
          "return": FiberNode {
            "_debugID": 3,
            "_debugIsCurrentlyTiming": false,
            "_debugOwner": null,
            "_debugSource": null,
            "alternate": null,
            "child": [Circular],
            "effectTag": 1,
            "firstEffect": null,
            "index": 0,
            "internalContextTag": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": Object {
              "Component": [Function],
              "context": null,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "memoizedState": Object {
              "context": null,
              "mount": true,
              "props": Object {
                "children": [Function],
                "onToggle": [Function],
                "render": [Function],
              },
            },
            "nextEffect": null,
            "pendingProps": null,
            "pendingWorkPriority": 0,
            "ref": null,
            "return": FiberNode {
              "_debugID": 1,
              "_debugIsCurrentlyTiming": false,
              "_debugOwner": null,
              "_debugSource": null,
              "alternate": FiberNode {
                "_debugID": 1,
                "_debugIsCurrentlyTiming": false,
                "_debugOwner": null,
                "_debugSource": null,
                "alternate": [Circular],
                "child": null,
                "effectTag": 0,
                "firstEffect": null,
                "index": 0,
                "internalContextTag": 0,
                "key": null,
                "lastEffect": null,
                "memoizedProps": null,
                "memoizedState": null,
                "nextEffect": null,
                "pendingProps": null,
                "pendingWorkPriority": 1,
                "ref": null,
                "return": null,
                "sibling": null,
                "stateNode": Object {
                  "containerInfo": <div />,
                  "context": Object {},
                  "current": [Circular],
                  "isScheduled": false,
                  "nextScheduledRoot": null,
                  "pendingContext": null,
                },
                "tag": 3,
                "type": null,
                "updateQueue": Object {
                  "callbackList": null,
                  "first": Object {
                    "callback": null,
                    "isForced": false,
                    "isReplace": false,
                    "isTopLevelUnmount": false,
                    "next": null,
                    "partialState": Object {
                      "element": <WrapperComponent
                        Component={[Function]}
                        context={null}
                        props={
                          Object {
                            "children": [Function],
                            "onToggle": [Function],
                            "render": [Function],
                          }
                        }
                      />,
                    },
                    "priorityLevel": 1,
                  },
                  "hasForceUpdate": false,
                  "isProcessing": false,
                  "last": Object {
                    "callback": null,
                    "isForced": false,
                    "isReplace": false,
                    "isTopLevelUnmount": false,
                    "next": null,
                    "partialState": Object {
                      "element": <WrapperComponent
                        Component={[Function]}
                        context={null}
                        props={
                          Object {
                            "children": [Function],
                            "onToggle": [Function],
                            "render": [Function],
                          }
                        }
                      />,
                    },
                    "priorityLevel": 1,
                  },
                },
              },
              "child": [Circular],
              "effectTag": 0,
              "firstEffect": null,
              "index": 0,
              "internalContextTag": 0,
              "key": null,
              "lastEffect": null,
              "memoizedProps": null,
              "memoizedState": Object {
                "element": <WrapperComponent
                  Component={[Function]}
                  context={null}
                  props={
                    Object {
                      "children": [Function],
                      "onToggle": [Function],
                      "render": [Function],
                    }
                  }
                />,
              },
              "nextEffect": null,
              "pendingProps": null,
              "pendingWorkPriority": 0,
              "ref": null,
              "return": null,
              "sibling": null,
              "stateNode": Object {
                "containerInfo": <div />,
                "context": Object {},
                "current": [Circular],
                "isScheduled": false,
                "nextScheduledRoot": null,
                "pendingContext": null,
              },
              "tag": 3,
              "type": null,
              "updateQueue": null,
            },
            "sibling": null,
            "stateNode": WrapperComponent {
              "_reactInternalFiber": [Circular],
              "_reactInternalInstance": Object {},
              "context": Object {},
              "props": Object {
                "Component": [Function],
                "context": null,
                "props": Object {
                  "children": [Function],
                  "onToggle": [Function],
                  "render": [Function],
                },
              },
              "refs": Object {},
              "state": Object {
                "context": null,
                "mount": true,
                "props": Object {
                  "children": [Function],
                  "onToggle": [Function],
                  "render": [Function],
                },
              },
              "updater": Object {
                "enqueueForceUpdate": [Function],
                "enqueueReplaceState": [Function],
                "enqueueSetState": [Function],
                "isMounted": [Function],
              },
            },
            "tag": 2,
            "type": [Function],
            "updateQueue": null,
          },
          "sibling": null,
          "stateNode": [Circular],
          "tag": 2,
          "type": [Function],
          "updateQueue": null,
        },
        "_reactInternalInstance": Object {},
        "context": Object {},
        "props": Object {
          "children": [Function],
          "onToggle": [Function],
          "render": [Function],
        },
        "refs": Object {},
        "state": Object {
          "on": false,
        },
        "toggle": [Function],
        "updater": Object {
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
        },
      },
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": [Function],
        "onToggle": [Function],
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
